mautic/mautic
# Adding new `{resubscribe_url}` token

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [Enhancement]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | https://github.com/mautic/user-documentation/pull/342
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes 



#### Description:

This PR is adding the `{resubscribe_url}` token into the email builder.

##### Reason for such token

The need came up when we identified a bug in the one-click unsubscribe header and wanted to make sure that the contacts that were unsubscribed really wanted to be unsubscribed. We were sending double-opt-out emails and for this we needed a single link that would re-subscribe the contacts easily. We want to have it in the core such campaigns can help others at some point.

![Screenshot 2024-10-21 at 15 59 26](https://github.com/user-attachments/assets/3910a219-74d9-4756-bed6-a590f8772220)

This route already exists and is visible in the default unsubscribe page:
![Screenshot 2024-07-04 at 12 50 09](https://github.com/acquia/mc-cs/assets/1235442/777d490e-b1ed-439f-8a3c-91c8d2628a14)
So if someone clicks on the resubscribe link that this new token generates, they get into this resubscribe page:
![Screenshot 2024-07-04 at 12 51 52](https://github.com/acquia/mc-cs/assets/1235442/c1a766a0-7e98-4071-9928-f59f55739d47)


If the Mautic instance has the Preference center enabled, the usubscribe link from the email as well as from the resubscribe page will show up the preference center form:
![Screenshot 2024-07-04 at 12 53 46](https://github.com/acquia/mc-cs/assets/1235442/8d11b661-7f7c-4135-b4cc-9a33b25d5d9a)

Just as an FYI what happens when the email stat record gets deleted from the database:

The unsubscribe link still works when the email_stat record is deleted as at that point Mautic is using just the email address in the URL. However, the resubscribe link doesn't have that so this is how it looks like when I click the link when the email_stat record was deleted:

![Screenshot 2024-07-04 at 14 46 10](https://github.com/acquia/mc-cs/assets/1235442/f5910ac1-c895-4b09-bfee-d229d98419a7)



#### Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Test that the new token isn't showing up in other places like landing page, SMS and wherever Mautic show tokens.
3. Test that the new token is available in the email builder and is being correctly replaced when an email is sent.
4. Send with a SMTP transport and with some tokenized transport as well (Sparkpost)
5. Keep special attention that sending to multiple contacts will result to unique resubscribe links for that particular contact. No matter if it's the first one or the last one.

#### Other areas of Mautic that may be affected by the change:
1. Just the email tokens

#### List deprecations along with the new alternative:
1. None



[//]: # ( As applicable: )
#### List of areas covered by the unit and/or functional tests:
1. Sending an email to 2 contacts and checking that the token was replaced with a URL and the hash is the same as in the unsubscribe link. Also ensuring that the hashes aren't the same in both emails.
